### PR TITLE
feat: impl JsonTypdef for Url (behind a feature flag)

### PR DESCRIPTION
**Pull request by tomkarw**

### CODE REVIEW

**Comment by uint** [comment]
This is something I need to think about. It's not exactly desirable for a library to add feature flags, and if we start implementing the trait for foreign types, I'm not sure where to draw the line.

**Comment by tomkarw** [comment]
I get your point. However, it would definitely help with adoption.

Could you release a version with just the ordered field output? That's the one crucial for my use case.

**Comment by uint** [comment]
> Could you release a version with just the ordered field output? That's the one crucial for my use case.

Yes. I'm currently traveling, but I'll try to get it done tomorrow.

**Comment by uint** [comment]
Okay, let's do it. I had a think and figured expecting `jtd-derive` users to use workarounds for popular types is hard to justify. I'll only agree to implementations for popular crates, which this is.

File: Cargo.toml [hunk 1]
```diff
@@ -12,11 +12,15 @@ documentation = "https://docs.rs/jtd-derive"
 categories = ["encoding"]
 keywords = ["schema", "codegen", "json"]
 
+[features]
+url = ["dep:url"]
+
```
Comment (uint) [inline Cargo.toml:17]:
```suggestion
```

I'm fairly sure this is unnecessary. [The feature should be implicitly created bc of `optional = true`.](https://doc.rust-lang.org/cargo/reference/features.html#optional-dependencies)

**Review by uint — APPROVED** [review]

File: Cargo.toml (hunk 1 above)
Comment (tomkarw) [reply Cargo.toml:17]: TIL

**Comment by tomkarw** [comment]
I think we should do it for all/most of structs in `url` i.e. `Host`, `Origin`, etc? I work on it today.

**Comment by uint** [comment]
> I think we should do it for all/most of structs in url i.e. Host, Origin, etc? I work on it today.

`schemars` only implements the trait for `Url` - [see here](https://github.com/GREsau/schemars/blob/master/schemars/src/json_schema_impls/url.rs). I'm okay with adding others that are likely to be useful for JSON-based APIs and such.

(sorry about the edit - accident)

**Comment by uint** [comment]
BTW, any clues about why `minimal_deps` is failing? I'm scratching my head there.

**Comment by tomkarw** [comment]
Any pointers on how to manually implement `JsonTypedef` for an enum? I'm struggling to implement it for `url::Host` to be precise.

**Comment by tomkarw** [comment]
> BTW, any clues about why minimal_deps is failing? I'm scratching my head there.

That's a funny one. `cargo +nighlty update -Z minimal-versions` also downgraded `trybuild` to `1.0.49` and apparently there was a bug in that version, which made the tests that rely on `trybuild` fail. Bumping the minimal version helped. Out of curiosity, I might bisect the _40_ releases `trybuild` had to see what was the issue there.

**Comment by uint** [comment]
> Any pointers on how to manually implement `JsonTypedef` for an enum? I'm struggling to implement it for `url::Host` to be precise.

Ah, it looks like `Host` uses the default [externally tagged](https://serde.rs/enum-representations.html#externally-tagged) enum representation. JSON Typedef can't express that. I'm afraid we can't support it.

**Comment by tomkarw** [comment]
> I'm afraid we can't support it.

Alright. I'm happy to merge as-is, with just support for `Url`. Feel free to merge and release if you agree.

File: Cargo.toml [hunk 2]
```diff
@@ -17,6 +17,7 @@ jtd-derive-macros = { version = "=0.1.3", path = "macros" }
 serde = { version = "1.0.115", features = ["derive"] }
 serde_json = "1.0.50"
 thiserror = "1.0.3"
+url = { version = "2", optional = true }
 
 [dev-dependencies]
-trybuild = "1.0.49"
+trybuild = "1.0.89"
```
Comment (uint) [inline Cargo.toml:23]: Thanks for diagnosing this!

**Review by uint — APPROVED** [review]

**Comment by uint** [comment]
It's [all released](https://crates.io/crates/jtd-derive/0.1.4)!

**Comment by tomkarw** [comment]
Thanks for the quick dev cycle, it's not common in OSS 😅

**Pull request merged**